shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java

```diff
@@ -37,142 +49,282 @@
-        }
-        return super.onKeyUp(keyCode, event);
+  @Override
+  protected void onPause() {
+    super.onPause();
+    if (Keybase.appDidEnterBackground()) {
+      Keybase.appBeginBackgroundTaskNonblock(new KBPushNotifier(this, new Bundle()));
+    } else {
+      Keybase.setAppStateBackground();
     }
+  }
 
-    @Override
-    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
-        if (listener != null) {
-            listener.onRequestPermissionsResult(requestCode, permissions, grantResults);
-        }
-        if (permissions.length > 0 && permissions[0].equals("android.permission.READ_CONTACTS")) {
-            // Call callback wrapper with results
-            ContactsPermissionsWrapper.callbackWrapper(requestCode, permissions, grantResults);
+  private String readFileFromUri(ReactContext reactContext, Uri uri) {
+    if (uri == null) return null;
+
+    String filePath = null;
+    if (uri.getScheme().equals("content")) {
+      ContentResolver resolver = reactContext.getContentResolver();
+      String mimeType = resolver.getType(uri);
+      String extension = MimeTypeMap.getSingleton().getExtensionFromMimeType(mimeType);
+
+      // Load the filename from the resolver.
+      // Of course, Android makes this super clean and easy.
+      // Use a GUID default.
+      String filename = String.format("%s.%s", UUID.randomUUID().toString(), extension);
+      String[] nameProjection = {MediaStore.MediaColumns.DISPLAY_NAME};
+      Cursor cursor = resolver.query(uri, nameProjection, null, null, null);
+      if (cursor != null) {
+        try {
+          if (cursor.moveToFirst()) {
+            filename = cursor.getString(0);
+          }
+        } finally {
+          cursor.close();
         }
-        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
-    }
+      }
 
-    @Override
-    protected void onPause() {
-        super.onPause();
-        if (Keybase.appDidEnterBackground()) {
-            Keybase.appBeginBackgroundTaskNonblock(new KBPushNotifier(this));
-        } else {
-            Keybase.setAppStateBackground();
+      // Now load the file itself.
+      File file = new File(reactContext.getCacheDir(), filename);
+      try {
+        InputStream istream = resolver.openInputStream(uri);
+        OutputStream ostream = new FileOutputStream(file);
+
+        byte[] buf = new byte[64 * 1024];
+        int len;
+        while ((len = istream.read(buf)) != -1) {
+          ostream.write(buf, 0, len);
         }
+        filePath = file.getPath();
+      } catch (IOException ex) {
+        Log.w(TAG, "error writing shared file " + uri.toString());
+      }
+    } else {
+      filePath = uri.getPath();
     }
+    return filePath;
+  }
 
-    @Override
-    protected void onResume() {
-        super.onResume();
-        Keybase.setAppStateForeground();
-    }
+  // Emits the intent to JS
+  private void emitIntent(Intent intent) {
```

Comment:
Why was this moved into `MainActivity` instead of being a separate class as before? I think this makes it harder to read and understand what's going on. This method doesn't use anything in this class, so I don't see why it can't be separated out.

Reply:
Because this is only used by this class. That's why it's private. I can move it to it's own class but you don't get the same guarantees.